### ISSUE TITLE
Proxy for outgoing traffic when mirroring

### ISSUE THREAD

**Issue by Rydekull** [gogits/gogs#2594]
I\m having some slight trouble finding any configurable setting for proxy when doing outgoing/external traffic from GOGS, for example when trying to mirror content from github.

I've tried setting http_proxy, https_proxy and no_proxy but it doesn't seem like it's getting picked up.

Any suggestion or is this considered a RFE?

**Comment by Unknwon** [comment]
Hmm... Gogs currently does not have this feature, what would be the use case of this?

**Comment by Rydekull** [comment]
Well, exactly as I mentioned in the first post. But I can explain a bit further.

- - -

Imagine a scenario with the following:
You have a internal server on which you deploy gogs. This is to be used as your internal git server and all is well and all until you realize:
"Hey, I want some content from github appearing in my gogs automagically"
At which point you discover that there is a nifty mirror feature (https://github.com/gogits/gogs/issues/63) and your day is saved. You quickly login to gogs, click on mirror and are met with a nasty error saying:

"Migration failed: Clone: exist status 128 - fatal: unable to access 'https://github.com/gogits/gogs.git/': Failed to connect to github.com port 443: Connection refused"

You suddenly remember that your internal gogs server is running behind a firewall which does not allow any external traffic. Everything has to be going through a http-proxy.

- - -

Yeah, you probably get the idea. 

Normally this is solved by honoring the normal proxy environment variables http_proxy, https_proxy, no_proxy. And perhaps even adding local config parameters in the application, but at least following the environment variables.

See for example: 
https://www.gnu.org/software/wget/manual/html_node/Proxies.html

**Comment by Unknwon** [comment]
Thanks for the great detail!

> Normally this is solved by honoring the normal proxy environment variables http_proxy, https_proxy, no_proxy

Gogs currently calls `git clone --mirror xxx` for migration so I think the proxy you want to is this: http://stackoverflow.com/questions/783811/getting-git-to-work-with-a-proxy-server

Even if Gogs lets you set a config option, it still affects global.

**Comment by Unknwon** [comment]
ping @Rydekull

**Comment by Unknwon** [comment]
Close due to lack of feedback.

You can continue discussion on https://discuss.gogs.io/.

**Comment by hkbakke** [comment]
This actually seems to work in my setup. I did the following:

Added ```EnvironmentFile=/etc/default/gogs``` to my systemd service definition to have a portable way to define custom environment variables.

In that file I added:
<pre>
Use proxy when mirroring
http_proxy="http://proxy.example.com:3128"
https_proxy="http://proxy.example.com:3128"
no_proxy="127.0.0.1,127.0.1.1,localhost,.example.com"
</pre>

And restarted gogs:
```systemctl restart gogs.service```

And it worked. Not initially as expected though. git is kind of quirky in that it actually needs https_proxy defined for https repositories, where most software just uses http_proxy if https_proxy is not defined. 
Additionaly i initially had https_proxy="${http_proxy}" in this file as environment variables is normally sourced in a shell that substitutes the variables, but gogs complained that it could not understand the url "${http_proxy}" so I made an explisit definition.
